pkg/cmd/grafana-cli/commands/secretsmigrations/reencrypt_secrets.go

```diff
@@ -0,0 +1,207 @@
+package secretsmigrations
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/commands/runner"
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/utils"
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
+	"github.com/grafana/grafana/pkg/services/secrets"
+	"github.com/grafana/grafana/pkg/services/secrets/manager"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"xorm.io/xorm"
+)
+
+type simpleSecret struct {
+	tableName       string
+	columnName      string
+	isBase64Encoded bool
+}
+
+func (s simpleSecret) reencrypt(secretsSrv *manager.SecretsService, sess *xorm.Session) error {
+	var rows []struct {
+		Id     int
+		Secret string
+	}
+
+	selectSQL := fmt.Sprintf("SELECT id, %s as secret FROM %s", s.columnName, s.tableName)
+	if err := sess.SQL(selectSQL).Find(&rows); err != nil {
+		return err
+	}
+
+	for _, row := range rows {
+		if len(row.Secret) == 0 {
+			continue
+		}
+
+		var (
+			err     error
+			decoded = []byte(row.Secret)
+		)
+
+		if s.isBase64Encoded {
+			decoded, err = base64.StdEncoding.DecodeString(row.Secret)
+			if err != nil {
+				return err
+			}
+		}
+
+		decrypted, err := secretsSrv.Decrypt(context.Background(), decoded)
+		if err != nil {
+			return err
+		}
+
+		encrypted, err := secretsSrv.EncryptWithDBSession(context.Background(), decrypted, secrets.WithoutScope(), sess)
+		if err != nil {
+			return err
+		}
+
+		encoded := string(encrypted)
+		if s.isBase64Encoded {
+			encoded = base64.StdEncoding.EncodeToString(encrypted)
+		}
+
+		updateSQL := fmt.Sprintf("UPDATE %s SET %s = ? WHERE id = ?", s.tableName, s.columnName)
+		if _, err := sess.Exec(updateSQL, encoded, row.Id); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Column %s from %s has been re-encrypted successfully\n", s.columnName, s.tableName)
+
+	return nil
+}
+
+type jsonSecret struct {
+	tableName string
+}
+
+func (s jsonSecret) reencrypt(secretsSrv *manager.SecretsService, sess *xorm.Session) error {
+	var rows []struct {
+		Id             int
+		SecureJsonData map[string][]byte
+	}
+
+	selectSQL := fmt.Sprintf("SELECT id, secure_json_data FROM %s", s.tableName)
+	if err := sess.SQL(selectSQL).Find(&rows); err != nil {
+		return err
+	}
+
+	for _, row := range rows {
+		if len(row.SecureJsonData) == 0 {
+			continue
+		}
+
+		decrypted, err := secretsSrv.DecryptJsonData(context.Background(), row.SecureJsonData)
+		if err != nil {
+			return err
+		}
+
+		var toUpdate struct {
+			SecureJsonData map[string][]byte
+		}
+
+		toUpdate.SecureJsonData, err = secretsSrv.EncryptJsonDataWithDBSession(context.Background(), decrypted, secrets.WithoutScope(), sess)
+		if err != nil {
+			return err
+		}
+
+		if _, err := sess.Table(s.tableName).Update(toUpdate); err != nil {
```

Comment:
won't this update all rows in the table unless we specify a Where clause?

Reply:
Excellent! I guess that's why PR reviews exists 🙈 If you look into previous commits, the `ID` function call was there but as I experienced some weird errors with that one, for some reason I removed it in one of the refactors and not added back the `Where` clause.

Thanks!! 🙏🏻